src/pybullet_industrial/g_code_processor.py

```diff
@@ -0,0 +1,405 @@
+
+        elif g_cmd_type == 18:
+            self.axis = 1  # X-Z Axis
+
+        elif g_cmd_type == 19:
+            self.axis = 0  # Y-Z Axis
+
+        elif g_cmd_type in [0, 1, 2, 3]:
+            variables = {'X': np.nan, 'Y': np.nan, 'Z': np.nan, 'A': np.nan,
+                         'B': np.nan, 'C': np.nan, 'R': np.nan}
+
+            for val in cmd:
+                if val[0] in variables:
+                    variables[val[0]] = val[1]
+
+            xyz_val = np.array([variables['X'], variables['Y'],
+                                variables['Z']])
+            abc_val = np.array([variables['A'], variables['B'],
+                                variables['C']])
+            r_val = variables['R']
+
+            # Setting the new point considering offset
+            self.new_point = np.array([0.0, 0.0, 0.0])
+            for i, value in enumerate(xyz_val):
+                if np.isnan(value):
+                    self.new_point[i] = self.last_point[i]
+                else:
+                    self.new_point[i] = value + self.offset[0][i]
+
+            # Setting the new orientation considering offset
+            self.new_or = np.array([0.0, 0.0, 0.0])
+            for i, value in enumerate(abc_val):
+                if np.isnan(value):
+                    self.new_or[i] = self.last_or[i]
+                else:
+                    self.new_or[i] = value + self.offset[1][i]
+
+            # Moving the endeffector if there is a G0 Interpolation
+            if g_cmd_type == 0:
+                self.g0_interpolation()
+
+            # Moving the endeffector if ther is a G1 Interpolation
+            elif g_cmd_type in [1, 2, 3]:
+                self.g123_interpolation(g_cmd_type, r_val)
+
+    def g0_interpolation(self):
+        """The G0-interpolation sets the position of the endeffector
+        or tool without running an interpolation command
+
+        Args:
+            None
+        Returns:
+            None
+        """
+        orientation = p.getQuaternionFromEuler(self.new_or)
+        actv = self.active_endeffector  # abbreviation
+
+        if self.active_endeffector == -1:
+            for _ in range(20):
+                self.robot.set_endeffector_pose(self.new_point, orientation)
+                for _ in range(10):
+                    p.stepSimulation()
+                    time.sleep(self.sleep)
+
+                current_position = self.robot.get_endeffector_pose()[0]
+                or_euler = p.getEulerFromQuaternion(
+                    self.robot.get_endeffector_pose()[1])
+                current_orientation = np.array(or_euler)
+
+                position_error = np.linalg.norm(current_position -
+                                                self.new_point)
+                orientation_error = np.linalg.norm(current_orientation -
+                                                   self.new_or)
+
+                if position_error < 0.02 and orientation_error < 0.004:
+                    break
```

Comment:
[Comment 1] old gcode_class: "Consider moving the step simulation function directly in the run_gcode function. you currently have declared it in a number of places making it hard to track.

Ideally we would even want it outside of the run_gcode function as this current implementation prevents two robots from moving simulateniously using g code (as each calls its own step simulation)
@MaHajo we should proably talk about this"

[Comment 2] It would probably be best if the run_gcode function returns a iterator that performs the loop. This would make everything much more modular and usable

@liquidcronos @MaHajo: For now, I have created several methods inside the GCodeProcessor to ensure that the robot moves to the right positions. However, I am aware that this is not an optimal solution as similar methods are scattered throughout the codebase. I agree that it would be better to have the actual settings for running the simulation outside of the GCodeProcessor. Therefore, it would be beneficial to have a meeting to discuss this topic."